lib/rucio/daemons/rsedecommissioner/throttling.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+# Copyright European Organization for Nuclear Research (CERN) since 2012
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Classes used for throttling RSE decommissioning."""
+
+import logging
+from collections.abc import Callable
+from typing import Any
+
+from rucio.db.sqla.constants import RuleState
+
+
+class DeletionLimitExceeded(Exception):
+    """An exception raised when the numbers of deleted rules or locks exceed a threshold."""
+
+    def __init__(self, what: str, limit: int, count: int) -> None:
+        self.what = what
+        self.limit = limit
+        self.count = count
+
+    def __str__(self) -> str:
+        return (f'Number of deleted {self.what}={self.count} reached'
+                f' the limit={self.limit}')
+
+
+class LimitChecker:
+    """RSE decommissioning throttling agent.
+
+    A LimitChecker object holds counters and preset threshold values for the number of deleted
+    rules and locks. At each rule deletion operation, the `increment()` method should be called
+    with the deleted rule object to increment the rules counter by one and the locks counter by
+    the number of locks the rule claims to create. If the value of one of the counters exceeds
+    the corresponding threshold, a DeletionLimitExceeded exception is raised to notify the
+    calling program, at which point the program is expected to immediately halt the
+    decommissioning operation.
+    """
+
+    def __init__(
+        self,
+        max_rules: int,
+        max_locks: int
+    ) -> None:
+        self.max_rules = max_rules
+        self.max_locks = max_locks
+
+        self.deleted_rules = 0
+        self.deleted_locks = 0
+
+        self.processed_rules = 0
```

Comment:
I’m inclined to suggest that `processed_rules` shouldn’t be part of this class. It doesn’t affect the throttling in any way. It’s used by the profile to make decision as to whether the decommissioning is done.

Reply:
Please see the comment above; I figured that the LimitChecker class itself was not needed any more.